---
title: 'Lab 2: implementing marginal regression'
author: "Prof. Kelsey Grinde"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = TRUE)
```

## Install and Load Packages

Run the following code in the Console to install a new package called `snpstats` that we'll be using today:

    if (!require("BiocManager", quietly = TRUE))
        install.packages("BiocManager")
    BiocManager::install("snpStats")

Now, load this package along with a few others that will be helpful for our analysis. (You may need to install some of these other packages as well, especially if you've updated your version of R/RStudio since you took STAT 155.)

```{r load-packages, message = F}
library(snpStats)
library(dplyr)
library(ggplot2)
library(broom)
```

## Load Data

To download the data that we'll be using today, go to [this GitHub page](https://github.com/MareesAT/GWA_tutorial/) and download the `1_QC_GWAS.zip` file. Once you've downloaded that file successfully, you'll need to unzip it. On most computers, double-clicking the zipped file will unzip it, but you can also do it from the command line if that doesn't work. (Kelsey can help with this --- just raise your hand if you get stuck here!)

Once you've finished downloading the data, update the strings below to reflect the correct path to the data on your computer. Here's what it looks like for me (you'll need to update this):

```{r file-paths}
fam <- './1_QC_GWAS/HapMap_3_r3_1.fam'
bim <- './1_QC_GWAS/HapMap_3_r3_1.bim'
bed <- './1_QC_GWAS/HapMap_3_r3_1.bed'
```

Now, we can use the `read.plink` function in the `snpstats` package to read this data into R. (It might take a minute.)

```{r read-data, cache=TRUE}
hapmap <- read.plink(bed, bim, fam)
```

## Explore Data

The new `hapmap` data object is a list of three items:

```{r explore-data}
names(hapmap)
```

-   `genotypes` contains the genotype data
-   `fam` contains information on each individual
-   `map` contains information on each SNP

### Explore genotype data

Get info about the genotype data:

```{r explore-genotype-matrix}
hapmap$genotypes
```

How many individuals do we have in this dataset? How many SNPs?

> 165 individuals, and 11457897 snps

### Explore sample information

Look at the info we have on each individual:

```{r explore-sample-info}
head(hapmap$fam,20)
```

We'll ignore this information for now.

### Explore SNP information

We can also look at the info we have on each SNP:

```{r explore-snp-info}
head(hapmap$map)
```

What do you think each of these columns are telling us? Discuss with your group.

> for example, in the first row it is has a snp id (snp.name) and it is on chromosome 1. It has two versions of alleles prevalent: T or C.

One useful piece of information about these SNPs that the `map` data frame does not already contain is the minor allele frequency (MAF) --- that is, the frequency of the minor allele in this dataset. Thankfully, the `snpstats` package provides some helpful code for quickly calculating the MAF for each SNP in the dataset:

```{r calculate-MAF}
maf <- col.summary(hapmap$genotypes)$MAF
head(maf)
```

Notice that some of these SNPs have a MAF of zero! This means that everyone in the dataset has the same genotype at these positions. We refer to these SNPs as *monomorphic* (mono meaning "one").

Let's add the minor allele frequencies to our `map` data frame. (This will come in handy later.)

```{r add-MAF-to-SNP-info}
# add new MAF variable 
hapmap$map <- hapmap$map %>%
  mutate(MAF = maf)

# look at SNP info again
head(hapmap$map)
```

## Reformat data for analysis

The `snpstats` package uses a unique format to store data. Currently, genotypes are coded as 01, 02, and 03 (with 00 representing missing values:

```{r look-at-genotypes}
# look at first five rows/columns
hapmap$genotypes@.Data[1:5,1:5]
```

Let's convert this to the 0, 1, 2 (and NA) format that we've talked about in class:

```{r reformat-genotypes}
# convert from "SnpMatrix" to "numeric"
X <- as(hapmap$genotypes, "numeric")

# look at first five rows/columns to confirm conversion looks ok
X[1:5, 1:5]
```

If the conversion was successful, you should now see a matrix of 0's, 1's, and 2's.

**Before you go on, check in with the others at your table. Was everyone able to get to this point successfully? Does anyone have any questions so far?**

## Simulate Trait

Let's simulate a trait that depends on the SNP known as *rs2476601*. Here's what we know about this SNP:

```{r look-at-causal-SNP}
hapmap$map %>%
  filter(snp.name == 'rs2476601')
```

We can learn more about this SNP by searching the `dbSNP` database: <https://www.ncbi.nlm.nih.gov/snp/>. Enter the rsID (the SNP's name) into the dbSNP search bar and see what you can learn. Summarize your findings here:

> "alleles A\>G,T": the most prevalent allele is A and then G or T. There are three possible variation on this snp.
>
> "Gene: PTPN22 ([Varview](https://www.ncbi.nlm.nih.gov/variation/view/?q=rs2476601)), AP4B1-AS1 ([Varview](https://www.ncbi.nlm.nih.gov/variation/view/?q=rs2476601))": these two are the genes related to this snp.
>
> "Clinical significance: benign, risk factor": it can be a benign risk factor
>
> "A=0.115/69 (NorthernSweden)": in northern sweden pop, the frequency of A is 0.115 out of 69 persons.

Now, let's create a quantitative trait `y` that depends on the genotype at this SNP plus some random noise:

```{r simulate-trait}
n <- nrow(X)
y <- X[,'rs2476601'] + rnorm(n, 0, 1)
head(y)
```

Since this is a simulation, we know by design that our trait `y` depends on SNP rs2476601 and nothing else. We can use this information to check how well our methods are working. We'll refer to rs2476601 as the *causal SNP* from here on out.

## Run GWAS

In a real genetic study, we wouldn't know the location of the true causal SNP(s) beforehand, so we'd need to run a genome-wide association study to see if we can figure out which variants are associated with the trait. As we've discussed in class, the method we'll use for GWAS is *marginal regression*.

### Getting started

To start, let's look at what happens when we run marginal regression on the first five SNPs in this dataset:

```{r fit-initial-models}
## fit models at first few SNPs
mod1 <- lm(y ~ X[,1])
mod2 <- lm(y ~ X[,2])
mod3 <- lm(y ~ X[,3])
mod4 <- lm(y ~ X[,4])
mod5 <- lm(y ~ X[,5])
```

```{r summarize-initial-models}
## look at the model summaries
tidy(mod1)
tidy(mod2)
tidy(mod3)
tidy(mod4)
tidy(mod5)
```

Do you notice anything weird about the 1st and 3rd SNPs? What do you think is going on here?

> 1st and 3rd snps are monomorphic. Therefore the the matrix X columns' are linearly dependent. That's why we can only estimate intercept but not beta1.

**Stop and discuss this question with your table before you move on.**

### Remove monomorphic SNPs

We're not going to get any useful information from any of these monomorphic SNPs, so let's remove them. We can remove these SNPs from the `map` data frame with a `filter` statement: fill in the blanks (\_\_\_) below and then run this code chunk.

```{r remove-mono-from-map}
# keep only those SNPs with MAF > 0
map.clean <- hapmap$map %>%
  filter(MAF > 0)
nrow(map.clean)
```

How many SNPs are left after filtering? How many did we remove?

> 1283751 left.

Now, let's remove these monomorphic SNPs from our genotype matrix (`X`). Since we're working with a matrix here instead of a data frame, we can't use the `filter` function. Here's one way to remove the columns corresponding to the monomorphic SNPs from `X`:

```{r remove-mono-from-genotypes}
# create vector of which SNPs have a MAF of 0
monomorphic <- which(maf == 0) 
head(monomorphic) 

# remove columns in the monomorphic vector
X.clean <- X[,-monomorphic]

#confirmed the ncol and nrow
ncol(X.clean)
nrow(X.clean)
```

**Confirm that the new "clean" genotype matrix has the correct number of rows and columns before you move on.**

### Analyze chromosome 1

Even after removing the monomorphic SNPs, we still have `r ncol(X, clean)` variants remaining. This might take awhile to analyze in R, so let's focus on just the SNPs on the first chromosome to start.

Run the code chunk below to make a list of which SNPs live on chromosome 1:

```{r find-chr1-snps}
chr1.snps <- which(map.clean$chromosome == 1)
head(chr1.snps)
length(chr1.snps)
```

Now, we're going to loop through each of the SNPs on chromosome 1, fitting a linear regression model at each SNP. For each model, we'll record the estimates (`betas`), standard errors (`ses`), test statistics (`tstats`) and p-values (`pvals`) for the coefficient of interest (the slope).

```{r run-gwas-chr1}
# set up empty vectors for storing results
betas <- c()
ses <- c()
tstats <- c()
pvals <- c()

# loop through chromosome 1 SNPs
for(i in chr1.snps){
  # print out occasional updates telling us what SNP we're analyzing
  if(i %% 5000 == 0) print(paste('Analyzing SNP', i)) 
  # fit model
  mod <- lm(y ~ X.clean[,i])
  # get coefficient information
  coefinfo <- tidy(mod)
  # record estimate, SE, test stat, and p-value
  betas[i] <- coefinfo$estimate[2]
  ses[i] <- coefinfo$std.error[2]
  tstats[i] <- coefinfo$statistic[2]
  pvals[i] <- coefinfo$p.value[2]
}
```

Let's add our results to our map data frame that contains information about each SNP:

```{r chr1-results}
# start with the map info for the chr 1 SNPs
chr1.results <- map.clean %>%
  filter(chromosome == 1)

# then add betas, SEs, etc.
chr1.results <- chr1.results %>%
  mutate(Estimate = betas,
         Std.Error = ses,
         Test.Statistic = tstats,
         P.Value = pvals)

# look at results
head(chr1.results)
```

Now we're ready to plot our results! We'd like to get a visual of which SNPs have small p-values (suggesting a possible association between the SNP and the trait). To start, create a scatterplot of the p-value of each SNP versus its position along the chromosome.

```{r plot-pvals}
ggplot(chr1.results, aes(x=position, y=P.Value))+geom_point()
```

What do you think? How well can you tell from this plot which SNPs have small p-values?

> The visualization is really bad here. The p-values are very small, making it hard to visualize.

One way to improve this visualization is to apply a log transformation to the p-values. When we have lots of numbers squished into a small range (0 to 1, in this case), a log transformation can be a nice way to spread these out. Try plotting the log (base 10) of the p-values for each SNP instead.

```{r plot-log-pvals}
ggplot(chr1.results, aes(x=position, y=log(P.Value)))+geom_point()
```

What do you notice now? Are the SNPs with small p-values the ones with small (more negative) values or large (closer to 0) values on the log base 10 scale?

> Much more clear view. SNPs with small p-values have small values (really negative ones!)

A traditional Manhattan plot uses a negative log transformation instead. Try plotting the $-log_{10}$ p-values now:

```{r plot-minus-log-pvals}
ggplot(chr1.results, aes(x=position, y=-log(P.Value)))+geom_point()
```

What similarities do you notice between this plot and the Manhattan plots we've seen before? What differences do you notice? Discuss with your table.

> Very similar! Now, the SNPs with small p-values are on the top of the plot (with greater -log(P-val).) I don't understand the discrepancy in the middle.

### Analyze all chromosomes

As time allows, try repeating the analysis above, but now looking at the SNPs on other chromosomes as well. *Hint: the main thing you'll need to change is which SNPs you're looping over in your for loop.*

```{r}
chr3.snps <- which(map.clean$chromosome == 3)
head(chr3.snps)
length(chr3.snps)

betas <- c()
ses <- c()
tstats <- c()
pvals <- c()

# loop through chromosome 3 SNPs
for(i in chr3.snps){
  # print out occasional updates telling us what SNP we're analyzing
  if(i %% 5000 == 0) print(paste('Analyzing SNP', i)) 
  # fit model
  mod <- lm(y ~ X.clean[,i])
  # get coefficient information
  coefinfo <- tidy(mod)
  # record estimate, SE, test stat, and p-value
  betas[i] <- coefinfo$estimate[2]
  ses[i] <- coefinfo$std.error[2]
  tstats[i] <- coefinfo$statistic[2]
  pvals[i] <- coefinfo$p.value[2]
}

chr3.results <- map.clean %>%
  filter(chromosome == 3)

ncol(chr3.results)

head(betas)

# then add betas, SEs, etc.
chr3.results <- chr3.results %>%
  mutate(Estimate = betas,
         Std.Error = ses,
         Test.Statistic = tstats,
         P.Value = pvals)

# look at results
head(chr3.results)
```

```{r}
ggplot(chr3.results, aes(x=position, y=-log(P.Value)))+geom_point()
```

## Speed up the computation

The approach we've taken in this lab to implement marginal regression works, but it's not particularly efficient. Can you think of any ways that we could speed this up? Brainstorm some ideas with your group.

> Make a for loop to loop through the snps by chromosome.

Note: this could make an interesting project topic!
